File: ports/lua/lib/sprite.lua
-- sprite.lua: Sprite CLI execution helpers — sx, push_file, push_dir.

local config = require("config")

local M = {}

local shell_quote = config.shell_quote

--- Build the sprite args fragment (e.g. "-s 'my-sprite'" or "").
local function sprite_args(sprite_name)
    if sprite_name and sprite_name ~= "" then
        return "-s " .. shell_quote(sprite_name)
    end
    return ""
end
M.sprite_args = sprite_args

--- Run a command inside the sprite via bash.
-- Returns the exit code (0 = success).
function M.sx(sprite_name, cmd, dry_run)
    if dry_run then
        print('  [dry-run] sprite exec -s ' .. sprite_name .. ' bash -c "' .. cmd .. '"')
        return 0
    end

    local escaped = shell_quote(cmd)
    local full_cmd = "sprite exec -s " .. shell_quote(sprite_name) .. " bash -c " .. escaped
    local ok, exit_type, code = os.execute(full_cmd)
    if ok then return 0 end
    return code or 1
end

--- Run a command inside the sprite and capture stdout.
-- Returns the captured output string (trimmed).
function M.sx_capture(sprite_name, cmd, dry_run)
    if dry_run then
        print('  [dry-run] sprite exec -s ' .. sprite_name .. ' bash -c "' .. cmd .. '"')
        return ""
    end

    local escaped = shell_quote(cmd)
    local full_cmd = "sprite exec -s " .. shell_quote(sprite_name) .. " bash -c " .. escaped .. " 2>/dev/null"
    local handle = io.popen(full_cmd, "r")
    if not handle then return "" end
    local output = handle:read("*a")
    handle:close()

    -- Trim trailing whitespace/newlines
    if output then
        output = output:match("^(.-)%s*$") or ""
    else
        output = ""
    end

    return output
end

--- Get the directory part of a path (everything up to and including the last /).
local function dirname(path)
    local dir = path:match("^(.*)/[^/]*$")
    if dir and dir ~= "" then return dir end
    if path:sub(1, 1) == "/" then return "/" end
    return "."
end
M.dirname = dirname

--- Get the base name of a path (the last component).
local function basename(path)
    return path:match("([^/]+)$") or path
end
M.basename = basename

--- Push a local file to the sprite.
function M.push_file(sprite_name, src, dest, dry_run)
    if dry_run then
        print("  [dry-run] push " .. src .. " -> sprite:" .. dest)
        return 0
    end

    local dest_dir = dirname(dest)
    local sa = sprite_args(sprite_name)

    -- Create remote directory
    local mkdir_cmd = "sprite exec " .. sa .. " bash -c " .. shell_quote("mkdir -p '" .. dest_dir .. "'")
    os.execute(mkdir_cmd)

    -- Pipe file contents to sprite
    local cat_cmd = "sprite exec " .. sa .. " bash -c " .. shell_quote("cat > '" .. dest .. "'") .. " < " .. shell_quote(src)
    local ok, exit_type, code = os.execute(cat_cmd)
    if ok then return 0 end
    return code or 1
end

--- Push a local directory to the sprite (used by launch — preserves directory name).
function M.push_dir(sprite_name, src, dest, dry_run)
    if dry_run then
        print("  [dry-run] push dir " .. src .. " -> sprite:" .. dest)
        return 0
    end

    local sa = sprite_args(sprite_name)

    -- Create remote directory
    local mkdir_cmd = "sprite exec " .. sa .. " bash -c " .. shell_quote("mkdir -p '" .. dest .. "'")
    os.execute(mkdir_cmd)

    -- tar pipe: local tar -> sprite tar extract
    local src_dir = dirname(src)
    local src_base = basename(src)
    local dest_parent = dirname(dest)

    local tar_cmd = "tar czf - -C " .. shell_quote(src_dir) .. " " .. shell_quote(src_base)
                 .. " | sprite exec " .. sa .. " bash -c "
                 .. shell_quote("tar xzf - -C '" .. dest_parent .. "'")

    local ok, exit_type, code = os.execute(tar_cmd)
    if ok then return 0 end
    return code or 1
end

--- Push a local directory to the sprite with --strip-components=1.
-- Used by the standalone push subcommand where remote path IS the target dir.
function M.push_dir_strip(sprite_name, src, dest, dry_run)
    if dry_run then
        print("  [dry-run] push dir " .. src .. " -> sprite:" .. dest)
        return 0
    end

    local sa = sprite_args(sprite_name)
    local src_dir = dirname(src)
    local src_base = basename(src)

    local tar_cmd = "tar czf - -C " .. shell_quote(src_dir) .. " " .. shell_quote(src_base)
                 .. " | sprite exec " .. sa .. " bash -c "
                 .. shell_quote("mkdir -p '" .. dest .. "' && tar xzf - -C '" .. dest .. "' --strip-components=1")

    local ok, exit_type, code = os.execute(tar_cmd)
    if ok then return 0 end
    return code or 1
end

return M

File: ports/haskell/src/Sprite.hs
-- | Subprocess wrapper for the sprite CLI.
module Sprite
    ( sx
    , sxCapture
    , sxPassthrough
    , pushFile
    , pushDir
    , spriteList
    , spriteExists
    , spriteArgs
    ) where

import System.Exit     (ExitCode(..))
import System.IO       (hClose, hFlush, hGetContents, hPutStr)
import System.Process  ( CreateProcess(..), StdStream(..), createProcess
                       , proc, waitForProcess, callProcess, readProcess )
import System.FilePath  (takeDirectory)

-- | Build the [-s, spriteName] args list, or empty if sprite name is empty.
spriteArgs :: String -> [String]
spriteArgs ""   = []
spriteArgs name = ["-s", name]

-- | Run a bash command inside a sprite. Output goes to stdout/stderr of the
--   calling process (passthrough). Dry-run just prints. Returns exit code.
sxPassthrough :: String -> String -> Bool -> IO ExitCode
sxPassthrough spriteName cmd dryRun
    | dryRun = do
        putStrLn $ "  [dry-run] sprite exec -s " ++ spriteName ++ " bash -c \"" ++ cmd ++ "\""
        return ExitSuccess
    | otherwise = do
        (_, _, _, ph) <- createProcess (proc "sprite" args)
        waitForProcess ph
  where
    args = ["exec", "-s", spriteName, "bash", "-c", cmd]

-- | Run a bash command inside a sprite, capturing stdout. Dry-run just prints.
sxCapture :: String -> String -> Bool -> IO String
sxCapture spriteName cmd dryRun
    | dryRun = do
        putStrLn $ "  [dry-run] sprite exec -s " ++ spriteName ++ " bash -c \"" ++ cmd ++ "\""
        return ""
    | otherwise = do
        output <- readProcess "sprite" args ""
        return (strip output)
  where
    args = ["exec", "-s", spriteName, "bash", "-c", cmd]
    strip = reverse . dropWhile (== '\n') . reverse

-- | Run a bash command inside a sprite. Convenience wrapper that discards
--   the exit code but still passes through output. For dry-run, prints the command.
sx :: String -> String -> Bool -> IO ()
sx spriteName cmd dryRun = do
    _ <- sxPassthrough spriteName cmd dryRun
    return ()

-- | Push a local file to a sprite.
--   Equivalent to:
--     sprite exec -s SPRITE bash -c "mkdir -p $(dirname DEST)"
--     sprite exec -s SPRITE bash -c "cat > DEST" < src
pushFile :: String -> String -> String -> Bool -> IO ()
pushFile spriteName src dest dryRun
    | dryRun = putStrLn $ "  [dry-run] push " ++ src ++ " -> sprite:" ++ dest
    | otherwise = do
        let destDir = takeDirectory dest
        -- Create destination directory
        callProcess "sprite"
            ["exec", "-s", spriteName, "bash", "-c", "mkdir -p '" ++ destDir ++ "'"]
        -- Read local file and pipe to cat on the sprite
        fileContents <- readFile src
        (Just hin, _, _, ph) <- createProcess
            (proc "sprite" ["exec", "-s", spriteName, "bash", "-c", "cat > '" ++ dest ++ "'"])
            { std_in = CreatePipe }
        hPutStr hin fileContents
        hFlush hin
        hClose hin
        _ <- waitForProcess ph
        return ()

-- | Push a local directory to a sprite via tar.
--   Equivalent to:
--     sprite exec -s SPRITE bash -c "mkdir -p DEST"
--     tar czf - -C parent base | sprite exec -s SPRITE bash -c "tar xzf - -C destParent"
pushDir :: String -> String -> String -> Bool -> IO ()
pushDir spriteName src dest dryRun
    | dryRun = putStrLn $ "  [dry-run] push dir " ++ src ++ " -> sprite:" ++ dest
    | otherwise = do
        let srcParent = takeDirectory src
            srcBase   = takeBaseName' src
            destParent = takeDirectory dest
        -- Create destination directory
        callProcess "sprite"
            ["exec", "-s", spriteName, "bash", "-c", "mkdir -p '" ++ dest ++ "'"]
        -- tar czf - -C parent base | sprite exec ... bash -c "tar xzf - -C destParent"
        (_, Just tarOut, _, tarPh) <- createProcess
            (proc "tar" ["czf", "-", "-C", srcParent, srcBase])
            { std_out = CreatePipe }
        (Just spriteIn, _, _, spritePh) <- createProcess
            (proc "sprite" [ "exec", "-s", spriteName, "bash", "-c"
                           , "tar xzf - -C '" ++ destParent ++ "'" ])
            { std_in = CreatePipe }
        -- Pipe tar output to sprite stdin
        tarContents <- hGetContents tarOut
        hPutStr spriteIn tarContents
        hFlush spriteIn
        hClose spriteIn
        _ <- waitForProcess spritePh
        _ <- waitForProcess tarPh
        return ()
  where
    -- Extract the basename from a path (last component).
    -- We avoid System.FilePath.takeBaseName because that strips extensions.
    takeBaseName' :: String -> String
    takeBaseName' p =
        let stripped = if not (null p) && last p == '/' then init p else p
        in reverse $ takeWhile (/= '/') (reverse stripped)

-- | Run 'sprite ls' and return stdout.
spriteList :: IO String
spriteList = readProcess "sprite" ["ls"] ""

-- | Check if a sprite with the given name already exists.
spriteExists :: String -> IO Bool
spriteExists spriteName = do
    output <- spriteList
    let lns = lines output
        -- Check for the sprite name as a whole word in each line
        hasName l = spriteName `elem` words l
    return $ any hasName lns
